src/main/scala/com/fulcrumgenomics/bam/pileup/ByCoordinatePileupBuilder.scala

```diff
@@ -0,0 +1,182 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2022 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam.pileup
+
+import com.fulcrumgenomics.bam.api.SamOrder.Coordinate
+import com.fulcrumgenomics.bam.api.{SamIterator, SamOrder, SamRecord, SamSource}
+import com.fulcrumgenomics.bam.pileup.ByCoordinatePileupBuilder.DefaultInitialCacheSize
+import com.fulcrumgenomics.commons.CommonsDef._
+import com.fulcrumgenomics.coord.LocatableUtil.GenomicOrdering
+import com.fulcrumgenomics.fasta.SequenceDictionary
+import htsjdk.samtools.util.{Interval, Locatable}
+
+import java.io.Closeable
+import scala.collection.mutable
+
+/** A class to lazily pileup SAM records for coordinate-maintaining or coordinate-advancing loci.
+  *
+  * @param records a by-name iterator of SAM records that is assumed to be coordinate-sorted.
+  * @param dict the sequence dictionary associated with the reference sequences.
+  * @param minBaseQ the minimum base quality that a base must have to contribute to a pileup.
+  * @param minMapQ the minimum mapping quality a read must have to contribute to a pileup.
+  * @param mappedPairsOnly if true, only allow read-pairs with both reads mapped to contribute to a pileup.
+  * @param includeDuplicates if true, allow reads flagged as duplicates to contribute to a pileup.
+  * @param includeSecondaryAlignments if true, allow reads flagged as secondary alignments to contribute to a pileup.
+  * @param includeSupplementalAlignments if true, allow reads flagged as supplementary alignments to contribute to a pileup.
+  * @param excludeMapPositionOutsideFrInsert if true, exclude any record of an FR pair where the site requested is outside the insert.
+  * @param initialCacheSize the initial size for the internal SAM record cache, set this to your expected pileup depth.
+  */
+class ByCoordinatePileupBuilder private(
+  records: => SamIterator,
```

Comment:
Referencing by-name is required so the iterator is not built unless absolutely needed since building this iterator will change the internal state of `SamSource` and its underlying HTSJDK `SamReader`. If this was no here many unit tests will fail, so regressions will be caught if anyone changes this.